command-expansion-server/src/utils/typeguards.ts

```diff
@@ -0,0 +1,19 @@
+/** Typeguard against a null value */
+export function isNotNull<T>(entity: T | null): entity is T {
+  return entity !== null;
+}
+
+/** Typeguard against an undefined value */
+export function isDefined<T>(entity: T | undefined): entity is T {
+  return entity !== undefined;
+}
+
+/** Typeguard resolved settled promises */
+export function isResolved<T>(promise: PromiseSettledResult<T>): promise is PromiseFulfilledResult<T> {
+  return promise.status === 'fulfilled';
+}
+
+/** Typeguard rejected settled promises */
+export function isRejected<T>(promise: PromiseSettledResult<T>): promise is PromiseRejectedResult {
+  return promise.status === 'rejected';
```

Comment:
Is there no `promise.isRejected()`? Does JS/TS use a string for `promise.status` instead of an enum?

Reply:
This is a javascript implementation thing. So no isRejected(), gotta just check status. There _might_ be an enum for it around somewhere, but I haven't checked if they have updated the standard libs with that or not

Reply:
Got it. That's odd, I'd expect JS to have used an enum here. Maybe an early version of JS used string and now for backwards compat. we're stuck with it.

Reply:
Well enums dont ACTUALLY exist in JS, they are an addition in TS that is actually just syntactic sugar for an object
```
enum PromiseState {
  Resolved = "resolved",
  Rejected = "rejected"
}
```

transpiles to

```
const PromiseState = {
  Resolved: 'resolved',
  Rejected: 'rejected'
}
```

Or in a simpler typed TS format equivalent to

```
interface PromiseState {
  Resolved: 'resolved',
  Rejected: 'rejected'
}
const PromiseState: PromiseState = {
  Resolved: 'resolved',
  Rejected: 'rejected'
}
```

Reply:
Ohh wow. Thanks, this is really good to know.